rfcs/syntax-string-interpolation.md

```diff
@@ -0,0 +1,122 @@
+# String interpolation
+
+## Summary
+
+New string interpolation syntax, and extending `string.format` to accept any arbitrary values without being exact about its types.
+
+## Motivation
+
+The problems with `string.format` are many.
+
+1. Must be exact about the types and its corresponding value.
+2. Using `%d` is the idiomatic default for most people, but this loses precision.
+   * `%d` casts the number into `long long`, which has a lower max value than `double` and does not support decimals.
+   * `%f` by default will format to the millionths, e.g. `5.5` is `5.500000`.
+   * `%g` by default will format up to the hundred thousandths, e.g. `5.5` is `5.5` and `5.5312389` is `5.53123`. It will also convert the number to scientific notation when it encounters a number equal to or greater than 10^6.
+   * To not lose precision, you need to use `%s`, but even so the type checker assumes you actually wanted strings.
+3. No support for `boolean`. You must use `%s` **and** call `tostring`.
+4. No support for values implementing the `__tostring` metamethod.
+5. Having to use parentheses around string literals just to call a method of it.
+
+## Design
+
+To fix issue #5, we will need new syntax in order to not change the meaning of existing strings. There are a few components of this new expression:
+
+1. A string chunk (`` `...{ ``, `}...{`, and `` }...` ``) where `...` is a range of 0 to many characters.
+   * `%` escapes `%` and `{`, whereas `\` escapes `` ` ``.
+   * Restriction: the string interpolation literal must have at least one value to interpolate. We do not need 3 ways to express a single line string literal.
+2. An expression between the braces. This is the value that will be interpolated into the string.
+3. Formatting style may follow the expression, delimited by `,`.
+   * Restriction: the formatting style must be constant at parse time.
+   * The formatting style will follow the same syntax as pre-existing syntax in `string.format`.
+
+To put the above into formal EBNF grammar:
+
+```
+fmtflags ::= `-' | `+' | ` ' | `#' | `0'
+fmtwidth ::= <NUMBER> [<NUMBER>]
+fmtprecision ::= `.' <NUMBER> [<NUMBER>]
+fmtspecifier ::= `c' | `d' | `i' | `e' | `E' | `f' | `g' | `G' | `o' | `q' | `s' | `u' | `x' | `X'
+
+interpfmt ::= [fmtflags][fmtwidth][fmtprecision][fmtspecifier]
```

Comment:
Can only one format flag be provided, or was this supposed to be `{fmtflags}`?

Reply:
Just one, this should've been called `fmtflag`. This is gone in the updated RFC.